File: frontend/src/services/computers.ts
import api from './api';

export interface Computer {
  id: number;
  name: string;
  distinguished_name: string;
  ip_address: string | null;
  operating_system: string | null;
  os_version: string | null;
  description: string | null;
  status: 'active' | 'disabled';
  site: string | null;
  department: string | null;
  days_since_logon: number | null;
  last_logon_timestamp: string | null;
  created_at: string;
  updated_at: string;
}

export async function listComputers(params?: Record<string, unknown>) {
  const { data } = await api.get('/computers/', { params });
  return data;
}

export async function getComputer(id: number): Promise<Computer> {
  const { data } = await api.get(`/computers/${id}`);
  return data;
}

export async function createComputer(body: Partial<Computer>): Promise<Computer> {
  const { data } = await api.post('/computers/', body);
  return data;
}

export async function updateComputer(id: number, body: Partial<Computer>): Promise<Computer> {
  const { data } = await api.put(`/computers/${id}`, body);
  return data;
}

export async function deleteComputer(id: number): Promise<void> {
  await api.delete(`/computers/${id}`);
}

export interface ComputerFilterOptions {
  operating_systems: string[];
  departments: string[];
  sites: string[];
}

export async function getComputerFilterOptions(): Promise<ComputerFilterOptions> {
  const { data } = await api.get('/computers/filter-options');
  return data;
}

File: backend/app/routers/computers.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from app.database import get_db
from app.core.dependencies import get_current_user
from app.services.computer_service import ComputerService
from app.schemas.computer import ComputerCreate, ComputerUpdate, ComputerResponse
from app.core.pagination import PaginatedResponse, PaginationParams

router = APIRouter(prefix="/api/computers", tags=["Computers"])


@router.get("/", response_model=PaginatedResponse[ComputerResponse])
def list_computers(
    params: PaginationParams = Depends(),
    search: str | None = Query(None, description="Search across name, IP, description"),
    status: str | None = Query(None, alias="status_filter", description="Filter by status: active|disabled"),
    department: str | None = Query(None, description="Filter by department OU in DN"),
    operating_system: str | None = Query(None, description="Filter by operating system"),
    stale: bool | None = Query(None, description="Filter: stale computers (90+ days no logon)"),
    sort_by: str = Query("name", description="Sort column"),
    sort_order: str = Query("asc", description="Sort direction: asc|desc"),
    db: Session = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    return ComputerService(db).list_computers(params, search, status, department, operating_system, stale, sort_by, sort_order)


@router.get("/filter-options")
def get_filter_options(
    db: Session = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    """Return distinct filter values for dropdowns."""
    return ComputerService(db).get_filter_options()


@router.get("/{computer_id}", response_model=ComputerResponse)
def get_computer(
    computer_id: int,
    db: Session = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    return ComputerService(db).get_computer(computer_id)


@router.post("/", response_model=ComputerResponse, status_code=201)
def create_computer(
    data: ComputerCreate,
    db: Session = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    return ComputerService(db).create_computer(data)


@router.put("/{computer_id}", response_model=ComputerResponse)
def update_computer(
    computer_id: int,
    data: ComputerUpdate,
    db: Session = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    return ComputerService(db).update_computer(computer_id, data)


@router.delete("/{computer_id}", status_code=204)
def delete_computer(
    computer_id: int,
    db: Session = Depends(get_db),
    _: dict = Depends(get_current_user),
):
    ComputerService(db).delete_computer(computer_id)
